examples/custom_rule.jl

```diff
@@ -0,0 +1,191 @@
+#     a runtime error will be thrown alerting the user to the missing activity rule rather than silently ignoring the rule."
+
+# Finally, it may be that either `x` or `y` are marked as [`Const`](@ref). We can in fact handle this case, 
+# along with the previous two cases, all together in a single rule:
+
+function forward(func::Const{typeof(f)}, RT::Type{<:Union{DuplicatedNoNeed, Duplicated}}, 
+                 y::Union{Const, Duplicated}, x::Union{Const, Duplicated})
+    println("Using our general custom rule!")
+    y.val .= x.val.^2 
+    if !(x <: Const) && !(y <: Const)
+        y.dval .= 2 .* x.val .* x.dval
+    elseif !(y <: Const) 
+        y.dval .= 0
+    end
+    if RT <: DuplicatedNoNeed
+        return sum(y.dval)
+    else
+        return Duplicated(sum(y.val), sum(y.dval))
+    end
+end
+
+# Note that there are also exist batched duplicated annotations for forward mode, namely [`BatchDuplicated`](@ref)
+# and [`BatchDuplicatedNoNeed`](@ref), which are not covered in this tutorial.
+
+# ## Defining a reverse-mode rule
+
+# Finally, let's look at how to write a simple reverse-mode rule! 
+# First, we write a method for [`EnzymeRules.augmented_primal`](@ref):
+
+function augmented_primal(config::ConfigWidth{1}, func::Const{typeof(f)}, ::Type{<:Active},
+                          y::Duplicated, x::Duplicated)
+    println("In custom augmented primal rule.")
+    if needs_primal(config)
+        return AugmentedReturn(func.val(y.val, x.val), nothing, nothing)
+    else
+        return AugmentedReturn(nothing, nothing, nothing)
+    end
+end
+
+# Let's unpack our signature for `augmented_primal` :
+# * We accepted a [`EnzymeRules.Config`](@ref) object with a specified width of 1, which means that our rule does not support batched reverse mode.
+# * Our function `f` was annotated [`Const`](@ref) as usual.
+# * We dispatched on an [`Active`](@ref) annotation for the return value. This is a special annotation for scalar values, such as our return value,
+#   that indicates that that we care about the value's derivative but we need not explicitly allocate a mutable shadow since it is a scalar value.
+# * We annotated `x` and `y` with [`Duplicated`](@ref) as in the forward mode case.
+
+# The body of our function simply checks if the `config` requires the primal, running it if so.
+# Our function returns an `AugmentedReturn` object specifying the return value, its shadow, and any extra tape information (none in this csae).
+# Note that we return a shadow of `nothing` since the return value is marked [`Active`](@ref).
+
+# Now, we write a method for [`EnzymeRules.reverse`](@ref):
+
+function reverse(config::ConfigWidth{1}, func::Const{typeof(f)}, out::Active, tape,
+                 y::Duplicated, x::Duplicated)
+    println("In custom reverse rule.")
+    y.dval .+= out.val 
+    x.dval .+= 2 .* x.val .* y.dval # accumulate into shadow, don't assign!
```

Comment:
is using `x.val` here incorrect, because it could have been mutated since? Do I need to use the tape if I want the rule to handle possible overwrites?

Reply:
It's correct if the modifiedBetween is false for that value, otherwise yeah you need the tape.

You can actually test this by modifying x in place after f in an overall function.

Reply:
For the array to put on the tape, do you recommend just allocating it inside `augmented_primal`, or doing something else?

Reply:
Yeah exactly that.